platforms/core-configuration/core-serialization-codecs/src/main/kotlin/org/gradle/internal/serialize/codecs/core/GroovyCodecs.kt

```diff
@@ -125,28 +140,46 @@ object ClosureCodec : Codec<Closure<*>> {
                 }
                 throw MissingPropertyException(propertyName)
             }
-            scriptReferenced()
+            scriptReferenced(invocationDescription = propertyName)
         }
 
         override fun setProperty(propertyName: String, newValue: Any?) {
             // See above for why this check happens
             if (targetMetadata.hasProperty(null, propertyName) == null) {
                 throw MissingPropertyException(propertyName)
             }
-            scriptReferenced()
+            scriptReferenced(invocationDescription = propertyName)
         }
 
         override fun invokeMethod(name: String, args: Any): Any {
             // See above for why this check happens
             if (targetMetadata.respondsTo(null, name).isEmpty()) {
-                throw MissingMethodException(name, Project::class.java, arrayOf())
+                throw MissingMethodException(name, targetMetadata.targetType, arrayOf())
             }
-            scriptReferenced()
+            scriptReferenced(invocationDescription = name)
         }
 
         private
-        fun scriptReferenced(): Nothing {
-            error("Cannot reference a Gradle script object from a Groovy closure as these are not supported with the configuration cache.")
+        fun scriptReferenced(invocationDescription: String): Nothing {
```

Comment:
💭  This is the main change: instead of throwing a naked exception, report a CC problem